feat(vim.pack): more actions in confirm-update

echasnovski opened the pull request:
This PR adds several interactive capabilities to a confirmation buffer.

### LSP code actions

Didn't intend to add them, but was curious to see how it fits in the overall "in-process LSP server" approach. This can also serve as a future reference of code actions for in-process LSP servers.

Currently implemented actions are (act on "plugin at cursor"):
  - "Update" and "Skip updating" (if has updates). The idea is the use case of looking at changelog and not wanting to update right away some plugins. So instead of "Yank hash(es), update `version` in 'init.lua', restart, `vim.pack.update()`, and possibly revert 'init.lua' updates" use code action.
  - "Delete". The idea is seeing a plugin that user forgot to delete. Instead of "yank plugin name and `:lua vim.pack.delete({ '<C-r>0' })<CR>` use code action.

After applying code action, plugin's section is removed from the update, so `:write` only acts on the left plugins.

### Jump to next/previous plugin section

Map locally to 'nvim-pack' buffer the `[[` and `]]` to jump to next/previous plugin section.

---

Here is a demo:

https://github.com/user-attachments/assets/ee426ea6-5687-46f6-8b27-f7527f135a6d

---

Resolve #34773

justinmk commented on runtime/lua/vim/pack.lua:
```diff
@@ -875,11 +903,11 @@ function M.update(names, opts)
-      { 'fetch', '--quiet', '--tags', '--force', '--recurse-submodules=yes', 'origin' },
-      p.path
-    )
+    if not opts.__offline then
```
why is double-underscore used here? single underscore means "private don't use". there is no other "super private" variation...

echasnovski replied:
There are cases of using double-underscore to mean "super private" and I remember to have seen that in Matrix discussions. Using single underscore is indeed easier, I agree.

justinmk commented on runtime/lua/vim/pack/_lsp.lua:
```diff
@@ -22,6 +23,48 @@ local get_confirm_bufnr = function(uri)
   return tonumber(uri:match('^nvim%-pack://(%d+)/confirm%-update$'))
 end
 
+local GROUP_HEADER_PATTERN = '^# (%S+)'
+local PLUGIN_HEADER_PATTERN = '^## (.+)$'
```
lowercase

echasnovski commented on runtime/lua/vim/pack.lua:
```diff
@@ -822,6 +822,31 @@ local function show_confirm_buf(lines, on_finish)
+--- confirmation buffer under "# Update" section.
+--- @param bufnr integer
+--- @return table<string,boolean>
+local function get_confirmed_update_plugins(bufnr)
```
It's just that using (what looks like) three verbs in a row throws me off.
Probably something like `get_update_map()` is better here.

justinmk commented:
> * Update" and "Skip updating" (if has updates). ... instead of "Yank hash(es), update `version` in 'init.lua' ... use code action.
> * "Delete". ... Instead of "yank plugin name and `:lua vim.pack.delete({ '<C-r>0' })<CR>` use code action.

Unless these update init.lua, this is unlikely to see much usage. Managing plugins for 1 session is a pretty niche thing for users to do.

I guess it could become more relevant if we support `pkg.json` or some other way to define the user's plugins instead of inlining in init.lua

echasnovski commented on runtime/doc/pack.txt:
```diff
@@ -364,14 +364,19 @@ update({names}, {opts})                                    *vim.pack.update()*
+        buffer-local mappings:
+        • |]]| and |[[| to navigate through plugin sections.
+        It has special in-process LSP server attached to provide more
+        interactive features. Currently supported methods:
```
What part do you mean as implementation detail here? The goal of this is to describe to users what extra interactive actions/features are available inside confirmation buffer. Those are:
- `[[` / `]]` navigate through plugin headers and not the built-in notion of `:h section`.
- There is in-process LSP server with certain set of features. Mentioning LSP is important because it allows users to be aware that their custom LSP mappings for document symbols, hover, and code actions do work here.

Do you suggest mention nothing at all? Or Leave currently supported methods of in-process LSP?

justinmk commented on runtime/doc/pack.txt:
```diff
@@ -370,14 +370,19 @@ update({names}, {opts})                                    *vim.pack.update()*
+        while the ones starting with `<` will be reverted. It has dedicated
+        buffer-local mappings:
+        • |]]| and |[[| to navigate through plugin sections.
+        It has special in-process LSP server attached to provide more
+        interactive features. Currently supported methods:
```
```suggestion
        More features are provided via LSP:
```

justinmk commented on runtime/ftplugin/nvim-pack.lua:
```diff
@@ -49,3 +50,31 @@ for i, l in ipairs(lines) do
+-- Mappings
+local jump = function(direction)
+  local lnum = vim.api.nvim_win_get_cursor(0)[1]
+  local buf_lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
+  local from, to, by = lnum + 1, #buf_lines, 1
+  if direction == 'prev' then
+    from, to, by = lnum - 1, 1, -1
+  end
+
+  for i = from, to, by do
+    if buf_lines[i]:match('^## ') then
+      vim.api.nvim_win_set_cursor(0, { i, 0 })
```
Does `vim.fn.search()` achieve this with less code?

echasnovski replied:
Maybe, yeah. If Vimscript functions are generally okay to use in such situations, I'll take a look.

echasnovski replied:
Yeah, it does simplify the code quite a bit :+1:  Will change it.

justinmk commented on runtime/lua/vim/pack/_lsp.lua:
```diff
@@ -22,6 +23,48 @@ local get_confirm_bufnr = function(uri)
+
+--- @return { group: string?, name: string?, from: integer?, to: integer? }
+local get_plug_data_at_lnum = function(bufnr, lnum)
+  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
+  --- @type string, string, integer, integer
+  local group, name, from, to
+  for i = lnum, 1, -1 do
+    group = group or lines[i]:match(group_header_pattern) --[[@as string]]
```
Does `vim.regex:match_line()` avoid a lot of this code? It can match a buffer line.

(We should add `nvim_buf_match()` but oh well, vim regex is fine :)

echasnovski replied:
I can't see now how it can remove *a lot* of code.

To get a full structural data about particular line number, it has to go up to see its nearest headings (second level for plugin name, first level for plugin group like "Error"/"Update"/"Same") and down to see where it ends.

`vim.regex:match_line()` would still require iterating over lines, the same as this approach. The only code save I see is that there would be no need to get all buffer lines.

The patterns for the name and group headings are also used later, so they are also not remove-able by refactoring this part.

justinmk reviewed (approved):
codeactions are a very high-leverage pattern for exposing a menu of actions in interactive plugin-owned buffers (aka, custom UI).

off-topic nit: instead of `nvim-pack://6/confirm-update` , based on the discussion in https://github.com/neovim/neovim/issues/3278 I think "slug" items like `6` should be in the URI "fragment": `nvim-pack://confirm-update#6`. That also has a nice "big-endian" feel.

echasnovski committed:
feat(pack): add `[[` and `]]` mappings in confirmation buffer

Problem: No easy/robust way to jump between plugin sections.

Solution: Add `[[` and `]]` mappings.

echasnovski committed:
feat(pack): add code actions in confirmation buffer

Problem: No way to granularly operate on plugins when inside
confirmation buffer.

Solution: Implement code actions for in-process LSP that act on "plugin
at cursor":
  - Update (if has updates).
  - Skip updating (if has updates).
  - Delete.

  Activate via default `gra` or `vim.lsp.buf.code_action()`.

justinmk commented on test/functional/plugin/pack_spec.lua:
```diff
@@ -1136,6 +1136,7 @@ describe('vim.pack', function()
 
       it('has in-process LSP features', function()
         t.skip(not is_jit(), "Non LuaJIT reports errors differently due to 'coxpcall'")
+        track_nvim_echo()
```
nit: is "spy" the usual name for this?
```suggestion
        spy_nvim_echo()
```

echasnovski replied:
Honestly first time hearing it :eyes:

justinmk commented on test/functional/plugin/pack_spec.lua:
```diff
@@ -1203,6 +1204,150 @@ describe('vim.pack', function()
+        end
+
+        -- - Should not include "namespace" header as "plugin at cursor"
+        validate_action({ 1, 1 }, {}, 0)
```
If these are test assertions let's always use `assert`. We are pretty consistent about that (in some places we use `expect`). And we are consistent about "validate" denoting parameter/input validation (not a test concept).
```suggestion
        assert_action({ 1, 1 }, {}, 0)
```

echasnovski replied:
I forget about that, sorry. Will have to do a cleaning pass on the whole `pack_spec.lua`.
